apps/api/src/service/collaboratorsService.ts

```diff
@@ -0,0 +1,92 @@
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import { type PostgresDb } from '@/db/index.js';
+import { collaborators } from '@/db/schemas/collaborators.js';
+import logger from '@/logger.js';
+import { type AsyncResult, failure, success } from '@/utils/results.js';
+import { and, eq } from 'drizzle-orm';
+import { type CollaboratorModel, type CollaboratorRecord } from './types.js';
+
+const collaboratorsSvc = (db: PostgresDb) => ({
+	createCollaborators: async ({
+		newCollaborators,
+	}: {
+		newCollaborators: CollaboratorModel[];
+	}): AsyncResult<CollaboratorRecord[]> => {
+		try {
+			// Check for Duplicates
+			let hasDuplicateCollaborators = false;
+
+			for await (const collaborator of newCollaborators) {
+				const existingCollaborator = await db
+					.select()
+					.from(collaborators)
+					.where(
+						and(
+							eq(collaborators.first_name, collaborator.first_name),
+							eq(collaborators.last_name, collaborator.last_name),
+							eq(collaborators.institutional_email, collaborator.institutional_email),
+							eq(collaborators.position_title, collaborator.position_title),
+							eq(collaborators.application_id, collaborator.application_id),
+						),
+					);
+
+				if (existingCollaborator.length > 0) hasDuplicateCollaborators = true;
+			}
+
+			if (hasDuplicateCollaborators) {
+				return failure(`Cannot create duplicate collaborator records`, 'DuplicateRecords');
+			}
+
+			// Create Collaborators
+			const collaboratorRecords = await db.transaction(async (transaction) => {
+				const newRecords: CollaboratorRecord[] = [];
+
+				newCollaborators.forEach(async (collaborator) => {
+					// TODO: Inserting multiple records as an array is not working despite Drizzle team saying the issue is resolved: https://github.com/drizzle-team/drizzle-orm/issues/2849
+					const newCollaboratorRecord = await transaction.insert(collaborators).values(collaborator).returning();
+
+					if (!newCollaboratorRecord[0]) {
+						throw new Error(`Error creating new collaborators: ${collaborator}`);
+					}
+
+					newRecords.push(newCollaboratorRecord[0]);
+				});
+
+				return newRecords;
+			});
+
+			if (!(collaboratorRecords.length === newCollaborators.length)) {
+				throw new Error(`Error creating new collaborators: ${collaboratorRecords}`);
+			}
```

Comment:
I'll update this with my Failure type changes in my auth PR... in the future, this is the purpose of the failure type and we can return that instead of throwing error. Failures can communicate expected error situations instead of escaping out into undocumented thrown exceptions.

Reply:
The thinking here is that DB actions can (and do) throw errors so it's necessary to wrap the whole block in a `try/catch`
So I funneled all failures into the `catch` block which always returns a `failure`, with error logging